tripal/tripal
# Add new third party settings needed by publish

# New Feature

### Issue #1924

### ~~Depends on #2001~~ (merged)

### Closes #1994

### Tripal Version: 4.x

## Description
This PR adds two new third party settings intended for use by the publish service.
These two settings are used for publishing content types where more than one bundle is based on a single chado base table. For example, both the "gene" and "mRNA" content types are derived from the `feature` table.
The third party settings allow publish to know where to find the term that defines the bundle, whether in the base table, or in a property table.


## Testing?
This PR isolates the described changes from other changes in the publish PR #1991 to simplify review.

Testing involves:
1. Build a docker on 4.x branch
2. Import all of the type collections
3. Switch to this branch and test the update hook `drush updatedb`
4. Verify that the settings have now been added
```
$bundles = ['genetic_map', 'QTL', 'sequence_variant', 'genetic_marker', 'phenotypic_marker',
            'gene', 'mrna', 'physical_map', 'genome_assembly', 'genome_annotation', 'genome_project',
            'germplasm', 'breeding_cross', 'germplasm_variety', 'ril'];
foreach ($bundles as $bundle) {
  $entity_type_manager = \Drupal::entityTypeManager();
  $entity_type = $entity_type_manager->getStorage('tripal_entity_type')->load($bundle);
  $entity_type_chado_base_table = $entity_type->getThirdPartySetting('tripal', 'chado_base_table');
  $entity_type_bundle_type_table = $entity_type->getThirdPartySetting('tripal', 'bundle_type_table');
  $entity_type_bundle_type_column = $entity_type->getThirdPartySetting('tripal', 'bundle_type_column');
  print "bundle=\"$bundle\"\n";
  print "  base_table=\"$entity_type_chado_base_table\"\n";
  print "  type_table=\"$entity_type_bundle_type_table\"\n";
  print "  type_column=\"$entity_type_bundle_type_column\"\n";
}
```
expected output
```
bundle="genetic_map"
  base_table="featuremap"
  type_table="featuremapprop"
  type_column="type_id"
bundle="QTL"
  base_table="feature"
  type_table="feature"
  type_column="type_id"
bundle="sequence_variant"
  base_table="feature"
  type_table="feature"
  type_column="type_id"
bundle="genetic_marker"
  base_table="feature"
  type_table="feature"
  type_column="type_id"
bundle="phenotypic_marker"
  base_table="feature"
  type_table="feature"
  type_column="type_id"
bundle="gene"
  base_table="feature"
  type_table="feature"
  type_column="type_id"
bundle="mrna"
  base_table="feature"
  type_table="feature"
  type_column="type_id"
bundle="physical_map"
  base_table="featuremap"
  type_table="featuremapprop"
  type_column="type_id"
bundle="genome_assembly"
  base_table="analysis"
  type_table="analysisprop"
  type_column="type_id"
bundle="genome_annotation"
  base_table="analysis"
  type_table="analysisprop"
  type_column="type_id"
bundle="genome_project"
  base_table="project"
  type_table="projectprop"
  type_column="type_id"
bundle="germplasm"
  base_table="stock"
  type_table="stock"
  type_column="type_id"
bundle="breeding_cross"
  base_table="stock"
  type_table="stock"
  type_column="type_id"
bundle="germplasm_variety"
  base_table="stock"
  type_table="stock"
  type_column="type_id"
bundle="ril"
  base_table="stock"
  type_table="stock"
  type_column="type_id"
```
5. And for good measure, a docker built on this branch should already include these settings.

### Test creating a new content type - adding type field manually
1. Creating a new content type. Tripal -> Page Structure -> "+ Add custom type"
2. I created a "Tripal Project" content type based on the project table.
3. Add a "Tripal Project Name" field using the "Chado String Field Type" field - this gets the base table populated.
4. Add a "Tripal Project Type" field using the "Chado Type Reference" field. Term should match the bundle term. You will want to check the new box "This term defines the bundle"
![issue1924_gui-term-defines-bundle](https://github.com/user-attachments/assets/54696450-d24b-4cb3-bc18-2972b3c9e337)

### Test creating a new content type - adding type field manually
1-3. Repeat the steps 1-3 above, but then
4. Discover new fields
5. Edit the type field and you can then check the  "This term defines the bundle" box and save.
